### DISCUSSION TITLE
Issue with @types/node: BufferEncoding type doesn't match NodeJS implementation

### DISCUSSION THREAD

**Discussion by edwardgalligan** [DefinitelyTyped/DefinitelyTyped#54942] [Issues with a @types package]
@types/node BufferEncoding requires values to be all-lowercase, but [NodeJS accepts uppercase values](https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/src/api/encoding.cc#L13)

Definition:

- https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/buffer.d.ts#L155

I could PR a fix but I'm not 100% sure what the best approach is. Listing all possible values (e.g. `uTf-8`) seems overkill; perhaps simply all-lowercase & all-uppercase would suffice as a compromise. I haven't see capitalized values used in the wild (e.g. `Utf-8`), so that seems uncommon enough to leave out, but I may be wrong.

**Comment by KapitanOczywisty** [comment]
Uppercase are just for compatibility, fast checks are done against lowercase and then there is slower case-insensitive check. Another problem is that you would have multiple names for one option in autocompletion. Best is to use one format and stick with that.

[Example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQgrgMwRATgUQHYGMD2ATASwwHMoBeKAcgEMBnLAgyqAHyrmAQA5m3KOEAWh6t2nAIwA2ADYRe7LLQBM8-osErRlAEZ0IkgCyrdtfQbgppq6dWBFxxotRQhVACwgAPSgCg+AQXpGVQBVThE+MKEIqiipWVDFTUj1ZKoYPUNVDNNDCystABlbe2ynF1UACS9fAIBlAGEASSbQgBUAMRjKEM7hdo6pQrRQhrq0nrGNbP86tCytGFn5gxCAJULVQv82poA5B0X9-zWATSq0AA1KAG4fHwQ4bDscDChgCFpgAAoIbHwiMQAFywRDIdD-QgkACUIIAbjgCHgoABvKAAXzu9w+X2+ACI4HjoUA)
![image](https://user-images.githubusercontent.com/44417092/128492914-e0847017-a75f-49d3-9c04-502265cf8362.png)

**Reply by edwardgalligan** [reply]
@KapitanOczywisty I'm all for being strict, but that would only apply for defining & writing the NodeJS implementation itself. That's not within the remit of a types package: the types should aim to be implementation-compatible.

e.g. this current type will break an existing conformant codebase it's added to. That's generally not correct behaviour.

**Reply by KapitanOczywisty** [reply]
There are plenty of examples where something is working in JS, but is on purpose omitted in TS definitions. For example some functions will skip argument on any falsy value, but in types only `undefined` is listed. TS is stricter than JS and one of TS goals is to also "fix" some inconsistencies which JS has.

> e.g. this current type will break an existing conformant codebase it's added to. That's generally not correct behaviour.

Almost no code will work out-of-the-box with any strictness enabled when migrating from JS to TS. And besides that, there is no good way to allow case-insensitive options: duplicated autocomplete would result in some angry users, and making all strings comparisons case-insensitive would break even more things.

**Reply by edwardgalligan** [reply]
@KapitanOczywisty this isn't about migrating from JS to TS, this is about putting these typings in place on an existing TS codebase.

> here is no good way to allow case-insensitive options

I understand this (as commented in my original Q), but covering common use-cases seems like it would be a reasonable compromise.

**Reply by KapitanOczywisty** [reply]
> this isn't about migrating from JS to TS, this is about putting these typings in place on an existing TS codebase.

One more time: If you write code without typings you cannot expect that it will work flawlessly with typings. Why someone even would do that? It's event worse argument than JS->TS migration.

> covering common use-cases seems like it would be a reasonable compromise.

No, it wouldn't. This would annoy a lot of people just because someone cannot use replace all.. Please don't make your problem someone's else problem.

**Comment by edwardgalligan** [comment]
> Please don't make your problem someone's else problem.

My "problem" was phrased as an open question looking for feedback. I really didn't expect to receive ad hominem attacks instead.

> This would annoy a lot of people

I presume you believe it would annoy you. That means we have two datapoints. Please don't assume your personal use-case is any more generally representative than mine.

> code without typings

This is nothing to do with code without typings (who writes TS without types?)

`DefinitelyTyped` is a deliberative repository: many typings are incomplete, many libs are without typings and many new libs are being written. This is a core interface of Node's built-in lib so the typing will be naturally propagated to 3rd-party libs that directly reference those built-ins.

Some examples:

1. Example of an incomplete (built-in) typing being updated https://github.com/DefinitelyTyped/DefinitelyTyped/commit/4b33816b517d9c0f80a13e8c2d167560a558d137
2. Example of a 3rd-party lib having refs built-in typing interfaces added to signatures https://github.com/DefinitelyTyped/DefinitelyTyped/commit/99fe2e5223f44928ff4ff62c5ac16626c3955c40

Both of these types of changes (especially the latter) can impact an existing (*fully typed*) codebase on package update.

> Just because someone cannot use replace all..

Replace all doesn't work so great on `node_modules`. And when the issue is in your package supply-chain, the fact that the cause of it is a typing deliberately going against the implementation really doesn't help with getting it fixed.

**Reply by KapitanOczywisty** [reply]
With these example this makes way more sense, so with that best would be to fix dependencies.
